linera-execution/src/runtime.rs

```diff
@@ -460,6 +461,45 @@ impl SyncRuntimeInternal<UserContractInstance> {
         self.pop_application();
         Ok(())
     }
+
+    /// Runs the service in a separate thread as an oracle.
+    fn run_service_oracle_query(
+        &mut self,
+        application_id: ApplicationId,
+        query: Vec<u8>,
+    ) -> Result<Vec<u8>, ExecutionError> {
+        let context = QueryContext {
+            chain_id: self.chain_id,
+            next_block_height: self.height,
+            local_time: self.local_time,
+        };
+        let sender = self.execution_state_sender.clone();
+
+        let txn_tracker = TransactionTracker::default()
+            .with_blobs(self.transaction_tracker.created_blobs().clone());
+        let mut service_runtime =
+            ServiceSyncRuntime::new_with_txn_tracker(sender, context, txn_tracker);
+
+        // TODO(#3533): Use the timeout to limit execution time.
+        let _timeout = self
+            .resource_controller
+            .remaining_service_oracle_execution_time()?;
+        let execution_start = Instant::now();
+        let result = service_runtime.run_query(application_id, query);
+
+        // Always track the execution time, irrespective to whether the service ran successfully or
+        // timed out
+        self.resource_controller
+            .track_service_oracle_execution(execution_start.elapsed())?;
```

Comment:
so the timeout is shared across all calls within the block? it's not per-call.

Reply:
That is correct.